src/resolvelib/resolvers.py

```diff
@@ -412,6 +412,13 @@ def resolve(self, requirements, max_rounds):
                 # Dead ends everywhere. Give up.
                 if not success:
                     raise ResolutionImpossible(self.state.backtrack_causes)
+                else:
+                    criterion = self.state.criteria[name]
+                    # Put newly-pinned candidate at the end. This is essential because
+                    # backtracking looks at this mapping to get the last pin.
+                    for candidate in criterion.candidates:
+                        self.state.mapping.pop(name, None)
+                        self.state.mapping[name] = candidate
```

Comment:
Since `name` is unchanged throughout this section, this loop essentially popping and setting different objects to the same key over and over. Which seems wrong to me…?